### PR TITLE
fast sha2

### PR DESCRIPTION
**Pull request by gdkrmr**
An optimized version of sha2_256, more than twice as fast as before. If there is interest, I can also implement this for sha2_512.

### CODE REVIEW

**Comment by gdkrmr** [comment]
sha2_512 added, the speed up is over 2x for large files

File: stdlib/SHA/src/sha2.jl [hunk 1]
```diff
@@ -1,134 +1,162 @@
...
-            d = c
-            c = b
-            b = a
-            a = UInt32(T1 + T2)
-        end
+macro R1_16(j, T)
+
+    ww = (:a, :b, :c, :d, :e, :f, :g, :h)
+
+    a = ww[((81 - j) % 8) + 1]
```
Comment (StefanKarpinski) [inline stdlib/SHA/src/sha2.jl]: You may want to verify that this generates fully optimized code without bounds checks and using a simple bitwise and operation for modular arithmetic.

Comment (Keno) [reply stdlib/SHA/src/sha2.jl]: This is in the compile time part of the macro, so it doesn't matter.

**Comment by gdkrmr** [comment]
all the constants for the SHA package are saved as vectors, will it make a difference to save them as immutable tuples?

**Comment by StefanKarpinski** [comment]
@staticfloat, care to review as the original author of the SHA package?

**Comment by StefanKarpinski** [comment]
I think this should become the definitive location for the SHA package unless we want to continue to sync them using `git subtree`.

**Comment by staticfloat** [comment]
I kind of assumed that the definitive place would be the externally-visible repo.  I don't have any protectiveness over SHA.jl, but it makes sense to me that we would want to keep the package development within the packages, rather than in a monolithic repo.  Especially if our eventual aim is to not use `git subtree `syncing in the future, but to somehow vendor in packages from external repositories into the stdlib (which I had thought was the eventual goal)

**Comment by gdkrmr** [comment]
I found it quite confusing and a little bit tedious to edit the code in the main julia repo, it took me a call to `@edit` to find out that I was not actually testing my edits.

Other things are: Pkg.test("SHA") is not possible and sourcing the test file does not work either. Single function can be tested by sourcing `SHA.jl`. After every edit I had to call `make` on the julia repo and rebuild the entire image.

That said it would probably make things more complicated to have an external repo, because I cannot install SHA like a normal package any more.

**Comment by KristofferC** [comment]
You can but you need something like `pushfirst!(LOAD_PATH, Pkg.dir)`. Yeah this is not ideal right now.

**Comment by gdkrmr** [comment]
> Fantastic, macros make manual unrolling so much nicer.

The macro facilities in Julia are really cool and the code is so much cleaner than all the c/c++ implementations. The speed gain does not come from the unrolling. but from changing

```
...
f(a, b)
a, b = b, a
f(a, b)
...
```
to
```
...
f(a, b)
f(b, a)
...
```

File: stdlib/SHA/src/sha2.jl [hunk 2]
```diff
@@ -1,134 +1,162 @@
...
+function transform!(context::Union{SHA2_224_CTX,SHA2_256_CTX})
+    pbuf = buffer_pointer(context)
+    # Initialize registers with the previous intermediate values (our state)
+    a, b, c, d, e, f, g, h = context.state
+
+    # Initial Rounds
+    @R1_16( 1, 256); @R1_16( 2, 256); @R1_16( 3, 256); @R1_16( 4, 256);
+    @R1_16( 5, 256); @R1_16( 6, 256); @R1_16( 7, 256); @R1_16( 8, 256);
+    @R1_16( 9, 256); @R1_16(10, 256); @R1_16(11, 256); @R1_16(12, 256);
+    @R1_16(13, 256); @R1_16(14, 256); @R1_16(15, 256); @R1_16(16, 256);
```
Comment (Keno) [inline stdlib/SHA/src/sha2.jl]: I'm always weary of lists of numbers like this. Maybe use another macro:
```
@static_unroll for i = 1:16
    @R1_16(i, 256)
end
```

Comment (gdkrmr) [reply stdlib/SHA/src/sha2.jl]: done, is there a `@static_unroll` macro?

**Comment by KristofferC** [comment]
In the long run, I think the setup should definitely be that stdlibs live in their own repos, are versioned like normal packages and we have a Project + Manifest that the package manager uses to pull in the stdlibs when Julia is built. This would allow more focused discussion / testing / benchmarking of the stdlibs. The reason why I have wanted to wait a little bit with that is simply that the package manager and the project/manifest formats are not stable enough yet to go full out. 

For this specific PR, I think making it towards the SHA.jl repo and then use a simple cherry-pick with subtree strategy is the best. If the OP doesn't want to do that, I can make a cherry pick PR in the reverse direction.

File: stdlib/SHA/src/sha2.jl (hunk 2 above)
Comment (gdkrmr) [reply stdlib/SHA/src/sha2.jl]: i did:
```julia
macro R_init(T)
    expr = :()
    for i in 1:16
        expr = :($expr; @R1_16($i, $T))
     end
    return esc(expr)
 end
```

**Comment by gdkrmr** [comment]
No problem, I can do the PR against SHA.jl if that is the easiest way for you. Let us just finish the discussion here.

**Comment by gdkrmr** [comment]
I made a PR to SHA.jl. https://github.com/staticfloat/SHA.jl/pull/57

**Pull request closed without merging**